st2reactor/st2reactor/garbage_collector/base.py

```diff
@@ -0,0 +1,158 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Garbage collection service which deletes old data from the database.
+"""
+
+import datetime
+
+import eventlet
+from eventlet.support import greenlets as greenlet
+from oslo_config import cfg
+
+from st2common import log as logging
+from st2common.constants.exit_codes import SUCCESS_EXIT_CODE
+from st2common.constants.exit_codes import FAILURE_EXIT_CODE
+from st2common.util import isotime
+from st2common.util.date import get_datetime_utc_now
+from st2reactor.garbage_collector.executions import purge_executions
+from st2reactor.garbage_collector.trigger_instances import purge_trigger_instances
+
+__all__ = [
+    'GarbageCollectorService'
+]
+
+# TODO: Make config option
+DEFAULT_COLLECT_INTERVAL = 300
+
+# Minimum value for TTL. If user supplies value lower than this, we will throw.
+MINIMUM_TTL_DAYS = 7
+
+LOG = logging.getLogger(__name__)
+
+
+class GarbageCollectorService(object):
+    def __init__(self, collection_interval=60):
+        """
+        :param collection_interval: How often to check database for old data and perform garbage
+               collection.
+        :type collection_interval: ``int``
+        """
+        self._collection_interval = collection_interval
+
+        self._action_executions_ttl = cfg.CONF.garbagecollector.action_executions_ttl
+        self._trigger_instances_ttl = cfg.CONF.garbagecollector.trigger_instances_ttl
+        self._validate_ttl_values()
+
+        self._running = True
+
+    def run(self):
+        try:
+            self._main_loop()
+        except greenlet.GreenletExit:
+            self._running = False
+            return SUCCESS_EXIT_CODE
+        except Exception:
+            self._running = False
+            return FAILURE_EXIT_CODE
+
+        return SUCCESS_EXIT_CODE
+
+    def shutdown(self):
+        self._running = False
+
+    def _main_loop(self):
+        while self._running:
+            self._perform_garbage_collection()
+
+            LOG.info('Sleeping for %s seconds before next garbage collection...' %
+                     (self._collection_interval))
+            eventlet.sleep(self._collection_interval)
+
+    def _validate_ttl_values(self):
+        """
+        Validate that a user has supplied reasonable TTL values.
+        """
+        if self._action_executions_ttl and self._action_executions_ttl < MINIMUM_TTL_DAYS:
+            raise ValueError('Minimum possible TTL in days is %s' % (MINIMUM_TTL_DAYS))
+
+        if self._trigger_instances_ttl and self._trigger_instances_ttl < MINIMUM_TTL_DAYS:
+            raise ValueError('Minimum possible TTL in days is %s' % (MINIMUM_TTL_DAYS))
+
+    def _perform_garbage_collection(self):
+        LOG.info('Performing garbage collection...')
+
+        # todo sleep between queries to avoid busy waiting
+        if self._action_executions_ttl >= MINIMUM_TTL_DAYS:
```

Comment:
Yeah, I decided to leave it as it is since it doesn't really add any value right now.

Collection of action executions and trigger instances intentionally doesn't run in parallel to avoid too much pressure on the database. But as you have said, when we add more things here (especially ones which need to be collected / rectified more often) and the service gets more busy, using green threads / thread pool might make more sense.